src/upgrader.ts

```diff
@@ -258,56 +260,70 @@ export class DefaultUpgrader extends EventEmitter<UpgraderEvents> implements Upg
 
     log('Starting the outbound connection upgrade')
 
+    // If the transport natively supports encryption, skip connection
+    // protector and encryption
+
     // Protect
     let protectedConn = maConn
-    const protector = this.components.getConnectionProtector()
+    if (!skipEncryption) {
+      const protector = this.components.getConnectionProtector()
 
-    if (protector != null) {
-      protectedConn = await protector.protect(maConn)
+      if (protector != null) {
+        protectedConn = await protector.protect(maConn)
+      }
```

Comment:
We still want to do connection protection here - it's use case is a little different to encrypting the whole channel, instead you're essentially making a private network on top of the encrypted connection because you'll need the PSK to communicate any data.

That is, I can support XYX connection encryption scheme (noise, etc) and connect to a remote node, but I also need to know the PSK to connect to a remote node that uses protected connections.

Reply:
This will break if we're using native muxing since there isn't a single underlying "connection". But I agree that we should do this even with native encryption. Maybe skip this if we have a custom stream muxer? fwiw on go-libp2p the QUIC transport doesn't support PSK.

Reply:
@achingbrain Please correct me if I have misunderstood here: For a connection that does not have native encryption, we use the `PreSharedKeyConnectionProtector` to wrap the plaintext stream. Then, the selected encryption scheme runs on top of the already encrypted stream. Is this done for nodes to limit which other nodes can connect to them?

As @MarcoPolo has mentioned, we do not have a single underlying stream in the webrtc spec. Instead, we create an ephemeral data channel for the Noise handshake. Subsequent streams only use the native DTLS encryption.

Reply:
In go-libp2p, we don't support PSK in QUIC or WebTransport.

Reply:
We probably can't support a PSK with native encryption without giving this more thought. I think this change is okay.

Reply:
See https://github.com/libp2p/js-libp2p-interfaces/pull/290#discussion_r988230911 - it's a small change but I think we need to split skipping connection protection out into it's own option - it makes the logic in the changes here easier to follow and also leaves the door open for transports that do their own encryption and muxing but also support PSK.